### ISSUE TITLE
Provide a command that lists all supported resources

### ISSUE THREAD

**Issue by hauleth** [hashicorp/terraform#17018]
This would allow simpler integration with editors. Right now the best way to achieve such data is to scrape website, which isn't ergonomic and is hard when one want to support multiple Terraform versions.

**Comment by apparentlymart** [comment]
Hi @hauleth! Thanks for this suggestion.

We are planning to introduce features to support text editor and IDE integrations, but we are holding off on that for the moment while we work on a revamp of the configuration language. The needs of that project have caused us to start implementing the underlying features to support such an interface, including support for listing resource types and obtaining their respective schemas, but we want to wait until these new features have stabilized before exposing this for external consumption since we may need to change the details before release.

**Comment by SMBurrows** [comment]
Hi @apparentlymart, 

This is also a feature I'm interested in. 

I understand that there are changes being made to the configuration language which needs to happen before work can start on this. 

However, Do you have a rough idea when it's scheduled for and is there anything I can do to help move it along?

**Comment by apparentlymart** [comment]
Once [0.12](https://www.hashicorp.com/blog/terraform-0-1-2-preview) is released we'll have the necessary building blocks in place to build something like this. Supporting powerful integrations with text editors and IDEs is a big goal of mine, so I'm eager to get this in as soon as possible but the groundwork must be there first.

I don't yet know exactly how this will look, but I expect we will add a new command to the `terraform` CLI which will print out the full schema for every provider that is requested by the current configuration _if_ `terraform init` has already been run to make sure those plugins are available. Prior to plugin installation the information isn't available, so autocomplete and other such functionality would need to be suspended until the user has got the necessary plugins installed.

Another part of this we're thinking about is a way for an editor to detect when it may need to re-run the command to update a cache or index. For that, we're thinking about placing a file at a specific location under `.terraform` that the editor can either watch for changes (using e.g. `inotify` on Linux) or can check the last modification time, and then Terraform will update that file each time the set of installed plugins changes.

Some details still to be figured out when we get there, but that's the high-level idea. Please let me know if that seems reasonable or if you think something is missing there, and then we'll look back at this issue when we're ready to implement.

**Comment by SMBurrows** [comment]
That would make sense. 

I would have thought it would have to be a few layers deep
Assuming `terraform init` has already been executed:

1. Provider Level - In a nutshell, listing the data on the [provider config page](https://www.terraform.io/docs/providers/aws/index.html)
2. Resources - Get a list of all the available resources|datasource for the project
3. Resource - Get a list of Arguments and Attributes for a resource. Perhaps with some of the additional data the schema's have such as required

**Comment by hauleth** [comment]
@apparentlymart alternative would be to implement Language Server either within Terraform itself or as an external tool bounded with Terraform.

**Comment by apparentlymart** [comment]
Hi @hauleth,

A language server is definitely something we've thought about, but seems like a lot more work than a one-shot schema export tool. Not to say that there can't _ever_ be a language server implementation, but I'd lean towards making incremental progress via some simple commands for now and then we can potentially use the learnings from that experience (including some further commands for common things like "what's at the cursor?" "where is the definition of this?", "rename symbol", etc) for a later effort to fold it all together into a robust language server.

**Comment by VladRassokhin** [comment]
I'll leave couple of links here:

* Old PR to support 'schema' #3508  While it's closed I do still update my branch to extract data about functions and provisioners.
* [Repo with metadata](https://github.com/VladRassokhin/terraform-metadata) from point 1 and metadata for all official providers. It also contains scripts used to generate it.

**Comment by apparentlymart** [comment]
Terraform v0.12 has a new command [`terraform providers schema`](https://www.terraform.io/docs/commands/providers/schema.html) which will allow exporting the schema of all providers used in the current configuration as JSON. The result gets large quickly with the more complex providers (`aws`, etc) so I expect editor extensions will want to run it asynchronously and cache the result. Monitoring the contents of the `.terraform/plugins` directory for changes is one way to detect that a cache update may be needed.

The `terraform providers schema` command is already available in [v0.12.0-beta2](https://www.hashicorp.com/blog/announcing-terraform-0-1-2-beta-2) and in the master branch ready to be included in the forthcoming final release, so I'm going to close this out.

I also did some initial prototyping for a language server in the [`f-langserver`](https://github.com/hashicorp/terraform/compare/f-langserver) branch. We don't plan to implement that fully right now due to priorities being elsewhere, but it does seem like a good end state to centralize the main text editor integration functionality in a single place, rather than re-implementing it for each editor. We'll open a new issue about the language server when we're ready to prototype a little more and come up with a full plan for it.
